Flaky test: TestClientSendsAGoAway

This is the first time I'm seeing it: 
https://github.com/grpc/grpc-go/actions/runs/8787418013/job/24112571687?pr=7156

Might be worth to check it out.

```
--- FAIL: Test (47.61s)
    --- FAIL: Test/ClientSendsAGoAway (10.00s)
        tlogger.go:116: INFO clientconn.go:305 [core] [Channel #472]Channel created  (t=+184.125µs)
        tlogger.go:116: INFO clientconn.go:1680 [core] [Channel #472]original dial target is: "127.0.0.1:34251"  (t=+212.779µs)
        tlogger.go:116: INFO clientconn.go:1685 [core] [Channel #472]dial target "127.0.0.1:34251" parse failed: parse "127.0.0.1:34251": first path segment in URL cannot contain colon  (t=+246.482µs)
        tlogger.go:116: INFO clientconn.go:1706 [core] [Channel #472]fallback to scheme "passthrough"  (t=+274.044µs)
        tlogger.go:116: INFO clientconn.go:1714 [core] [Channel #472]parsed dial target is: passthrough:///127.0.0.1:34251  (t=+299.482µs)
        tlogger.go:116: INFO clientconn.go:1837 [core] [Channel #472]Channel authority set to "127.0.0.1:34251"  (t=+325.33µs)
        tlogger.go:116: INFO resolver_wrapper.go:197 [core] [Channel #472]Resolver state updated: {
              "Addresses": [
                {
                  "Addr": "127.0.0.1:34251",
                  "ServerName": "",
                  "Attributes": null,
                  "BalancerAttributes": null,
                  "Metadata": null
                }
              ],
              "Endpoints": [
                {
                  "Addresses": [
                    {
                      "Addr": "127.0.0.1:34251",
                      "ServerName": "",
                      "Attributes": null,
                      "BalancerAttributes": null,
                      "Metadata": null
                    }
                  ],
                  "Attributes": null
                }
              ],
              "ServiceConfig": null,
              "Attributes": null
            } (resolver returned new addresses)  (t=+464.651µs)
        tlogger.go:116: INFO balancer_wrapper.go:103 [core] [Channel #472]Channel switches to new LB policy "pick_first"  (t=+513.843µs)
        tlogger.go:116: INFO clientconn.go:846 [core] [Channel #472 SubChannel #473]Subchannel created  (t=+569.808µs)
        tlogger.go:116: INFO clientconn.go:535 [core] [Channel #472]Channel Connectivity change to CONNECTING  (t=+882.825µs)
        tlogger.go:116: INFO clientconn.go:305 [core] [Channel #472]Channel exiting idle mode  (t=+914.[194](https://github.com/grpc/grpc-go/actions/runs/8787418013/job/24112571687?pr=7156#step:4:195)µs)
        tlogger.go:116: INFO clientconn.go:1207 [core] [Channel #472 SubChannel #473]Subchannel Connectivity change to CONNECTING  (t=+971.722µs)
        tlogger.go:116: INFO clientconn.go:1322 [core] [Channel #472 SubChannel #473]Subchannel picks a new address "127.0.0.1:34251" to connect  (t=+998.061µs)
        tlogger.go:116: INFO clientconn.go:535 [core] [Channel #472]Channel Connectivity change to SHUTDOWN  (t=+1.652745ms)
        tlogger.go:116: INFO resolver_wrapper.go:100 [core] [Channel #472]Closing the name resolver  (t=+1.674466ms)
        tlogger.go:116: INFO balancer_wrapper.go:135 [core] [Channel #472]ccBalancerWrapper: closing  (t=+1.694022ms)
        tlogger.go:116: INFO clientconn.go:1[207](https://github.com/grpc/grpc-go/actions/runs/8787418013/job/24112571687?pr=7156#step:4:208) [core] [Channel #472 SubChannel #473]Subchannel Connectivity change to SHUTDOWN  (t=+1.721133ms)
        tlogger.go:116: INFO clientconn.go:1553 [core] [Channel #472 SubChannel #473]Subchannel deleted  (t=+1.773004ms)
        tlogger.go:116: INFO clientconn.go:305 [core] [Channel #472]Channel deleted  (t=+1.793723ms)
        goaway_test.go:826: Context timed out
FAIL
```

--- a/test/goaway_test.go
+++ b/test/goaway_test.go
@@ -800,6 +800,7 @@ func (s) TestClientSendsAGoAway(t *testing.T) {
 		for {
 			f, err := ct.fr.ReadFrame()
 			if err != nil {
+				t.Logf("error reading frame: %v", err)
 				return
 			}
 			switch fr := f.(type) {
@@ -808,6 +809,7 @@ func (s) TestClientSendsAGoAway(t *testing.T) {
 				if fr.ErrCode == http2.ErrCodeNo {
 					t.Logf("GoAway received from client")
 					close(goAwayReceived)
+					return
 				}
 			default:
 				t.Errorf("server tester received unexpected frame type %T", f)
@@ -816,6 +818,7 @@ func (s) TestClientSendsAGoAway(t *testing.T) {
 			}
 		}
 	}()
+	cc.WaitForStateChange(ctx, connectivity.Connecting)
 	cc.Close()
 	defer ct.conn.Close()
 	select {

--- a/test/goaway_test.go
+++ b/test/goaway_test.go
@@ -800,7 +800,7 @@ func (s) TestClientSendsAGoAway(t *testing.T) {
 		for {
 			f, err := ct.fr.ReadFrame()
 			if err != nil {
-				t.Logf("error reading frame: %v", err)
+				errCh <- fmt.Errorf("error reading frame: %v", err)
 				return
 			}
 			switch fr := f.(type) {

--- a/test/goaway_test.go
+++ b/test/goaway_test.go
@@ -770,33 +770,16 @@ func (s) TestClientSendsAGoAway(t *testing.T) {
 	if err != nil {
 		t.Fatalf("error listening: %v", err)
 	}
-	ctCh := testutils.NewChannel()
+	defer lis.Close()
+	goAwayReceived := make(chan struct{})
+	errCh := make(chan error)
 	go func() {
 		conn, err := lis.Accept()
 		if err != nil {
 			t.Errorf("error in lis.Accept(): %v", err)
 		}
 		ct := newClientTester(t, conn)
-		ctCh.Send(ct)
-	}()
-	defer lis.Close()
-
-	cc, err := grpc.Dial(lis.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()))
-	if err != nil {
-		t.Fatalf("error dialing: %v", err)
-	}
-
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer cancel()
-
-	val, err := ctCh.Receive(ctx)
-	if err != nil {
-		t.Fatalf("timeout waiting for client transport (should be given after http2 creation)")
-	}
-	ct := val.(*clientTester)
-	goAwayReceived := make(chan struct{})
-	errCh := make(chan error)
-	go func() {
+		defer ct.conn.Close()
 		for {
 			f, err := ct.fr.ReadFrame()
 			if err != nil {
@@ -818,9 +801,18 @@ func (s) TestClientSendsAGoAway(t *testing.T) {
 			}
 		}
 	}()
-	cc.WaitForStateChange(ctx, connectivity.Connecting)
+
+	cc, err := grpc.NewClient(lis.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("error dialing: %v", err)
+	}
+	cc.Connect()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	testutils.AwaitState(ctx, t, cc, connectivity.Ready)
 	cc.Close()
-	defer ct.conn.Close()
 	select {
 	case <-goAwayReceived:
 	case err := <-errCh:

